vmware_guest: various customizations changes and fixes

dagwieers opened the pull request:
##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
 - Bugfix Pull Request

##### COMPONENT NAME
<!--- Name of the module/plugin/module/task -->
vmware_guest

##### ANSIBLE VERSION
<!--- Paste verbatim output from “ansible --version” between quotes below -->
v2.3

##### SUMMARY
<!--- Describe the change, including rationale and design decisions -->

<!---
If you are fixing an existing issue, please include "Fixes #nnn" in your
commit message and your description; but you should still explain what
the change does.
-->
Most of my queued changes were already implemented by @aperigault !

This was still open
- Typos
- Various fixes to dict.get() without quotes
- Add defaults to `fullname` and `orgname` (so they are no longer mandatory)
- ~~Add default to `joinworkgroup` to WORKGROUP (maybe not a good idea afterall ?)~~
- Add missing timezone implementation
- Remove the customize flag from the options
- Rename 'customizations' to 'customization' (cfr VMware docs and fora)

dagwieers committed:
vmware_guest: various changes and fixes

Most of my queued changes were already implemented by @aperigault !

This was still open
- Typos
- Various fixes to dict.get() without quotes
- Defaults to fullname and orgname (so they are no longer mandatory)
- Add missing timezone implementation
- Remove the customize flag from the options
- Rename 'customizations' to 'customization' (cfr VMware docs and fora)

dagwieers commented:
@nerzhul @aperigault @dav1x @nitzmahone @jhawkesworth @jctanner Feedback welcome

nerzhul commented on lib/ansible/modules/cloud/vmware/vmware_guest.py:
```diff
@@ -461,7 +461,7 @@ def create_nic(self, device_type, device_label, device_infos):
         if device_type == 'pcnet32':
             nic.device = vim.vm.device.VirtualPCNet32()
-        if device_type == 'vmxnet2':
+        elif device_type == 'vmxnet2':
```
oh tricky i didn't noticed this typo

aperigault commented on lib/ansible/modules/cloud/vmware/vmware_guest.py:
```diff
@@ -1315,7 +1295,7 @@ def deploy_vm(self):
         self.configure_disks(vm_obj=vm_obj)
         self.configure_network(vm_obj=vm_obj)
-        if self.should_deploy_from_template() and len(self.params['customize']) > 0:
+        if vm_obj is None or self.should_deploy_from_template():
```
I don't understand. With your change we attempt to customize a vm created from scratch ?
I think we can't deploy VM from template without use customization with this.

dagwieers replied:
@aperigault Good question, I thought it was possible to customize newly created VMs as well.
Is that not correct ?

nerzhul replied:
this condition is correct as vm_obj is none on vm creation from scratch. This case is handled using same condition in every self.configure_xx
why customization should be forbidden ? If i want to create a VM from scratch using a PXE boot and after your sysprep ?

aperigault replied:
Ok we can use sysprep with from scratch VM. But if we run customization in all cases why this condition ?
I think it should be : if we set customizations params : if len(self.params['customize']) > 0:

nerzhul replied:
in fact, you are wrong @aperigault because if you add that you will break the guest_map customizations which are handled in customize_vm and doesn't require customize variable namespace

aperigault replied:
Oh yes! I missed it! But there always is a problem. How can I simply disable customization ? We should add tests in customize_vm or split customize_vm in two function ?

dagwieers replied:
Well, I was wondering the same thing. It only affects the stuff that defaults to something.
Does it really matter ? Except maybe speed ?
I don't mind re-introducing `customize:`, default it to `customize: yes`, although I don't like the name.

dagwieers replied:
Or assume that an empty customizations means no customizations (like you did).
It's easy to add a single item if needed.

aperigault replied:
I prefer the second one. If you want to use customization (others than network customization) you need to set customization params.

dagwieers replied:
Ok, agreed ! And implemented !

aperigault commented on lib/ansible/modules/cloud/vmware/vmware_guest.py:
```diff
@@ -1035,11 +1005,11 @@ def customize_vm(self, vm_obj):
+                ident.identification.domainadmin = str(self.params['customizations'].get('domainadmin'))
+                ident.identification.domainadminpassword = str(self.params['customizations'].get('domainadminpassword'))
+                ident.identification.joindomain = str(self.params['customizations'].get('joindomain'))
+            else:
```
I thought about it and I would prefer to not set default value for this param. It could be better to explicitly enable joindomain or joinworkgroup.

dagwieers replied:
I agree (as put in the description). Yesterday I thought it was a good idea, but maybe people have it already configured correctly in the template and we don't want to mess with it. Will undo.

dagwieers commented:
Shippable fails on:
```
Error response from daemon: Get https://registry-1.docker.io/v2/drydock/u14pyt/manifests/prod: dial tcp 52.0.53.94:443: connect: network is unreachable
```

Maybe we need an ansibot command to collect these kind of Shippable errors for statistics/troubleshooting ?

dagwieers commented:
ready_for_review

dagwieers commented:
ready_for_shipit :-)

dagwieers commented:
Important rename from "customizations" to "customization" (cfr. VMware documentation and fora).

dagwieers committed:
Important fixes for idempotency and customization

- A password is mandatory for customization to work on Windows !
- An important fix for idempotency related to guestId
- Support all types of Windows guestId entries

nerzhul commented on lib/ansible/modules/cloud/vmware/vmware_guest.py:
```diff
@@ -161,7 +161,7 @@
           - "  joinworkgroup (string): Workgroup to join (Not compatible with C(joindomain), default: WORKGROUP)"
           - "  orgname (string): Organisation name (default: ACME)"
-          - "  password (string): Local administrator password"
+          - "  password (string): Local administrator password (mandatory !)"
```
please remove the "!"

nerzhul commented on lib/ansible/modules/cloud/vmware/vmware_guest.py:
```diff
@@ -996,10 +996,12 @@ def customize_vm(self, vm_obj):
                 ident.guiUnattended.password.value = str(self.params['customization']['password'])
                 ident.guiUnattended.password.plainText = True
+            else:
+                self.module.fail_json(msg="A non-blank 'password' entry is mandatory in the 'customization' section.")
```
Remove "a non-blank"

dagwieers commented:
ready_for_review

aperigault commented on lib/ansible/modules/cloud/vmware/vmware_guest.py:
```diff
@@ -1322,6 +1297,7 @@ def deploy_vm(self):
         self.configure_disks(vm_obj=vm_obj)
         self.configure_network(vm_obj=vm_obj)
         if self.should_deploy_from_template() and len(self.params['customizations']) > 0:
```
You missed to remove this line
We already talk about it but what is the goal of this part ?
(vm_obj is None or self.should_deploy_from_template())

dagwieers replied:
To be honest I wasn't sure what my original code was after a merge-glitch during rebase (as it wasn't changed in my production code). So I put back what was there originally, not what I changed it to previously...

dagwieers commented:
@nerzhul I don't understand why you wanted me to remove the "non-blank" from the statement though. I test both if the password was defined and is non-empty. A non-empty password will cause the customization to fail and the module will end with a vague message that there was an error with customization.

So it is important that the password is not empty, as much as it is important that it is provided.

aperigault commented on lib/ansible/modules/cloud/vmware/vmware_guest.py:
```diff
@@ -1321,7 +1296,8 @@ def deploy_vm(self):
         self.configure_network(vm_obj=vm_obj)
-        if self.should_deploy_from_template() and len(self.params['customizations']) > 0:
+
+        if vm_obj is None or self.should_deploy_from_teplate():
```
I ask again: When this condition will be skipped ?
I understand: When we deploy from scratch or from template. There is an other case ?

aperigault commented:
Can we merge that to fix #20038 ?

aperigault commented on lib/ansible/modules/cloud/vmware/vmware_guest.py:
```diff
@@ -1321,7 +1296,8 @@ def deploy_vm(self):
         self.configure_network(vm_obj=vm_obj)
-        if self.should_deploy_from_template() and len(self.params['customizations']) > 0:
+
+        if self.should_deploy_from_template() and len(self.params['customization']) > 0:
```
With this we can't customize from scratch vm. I think we only need:
if len(self.params['customization']) > 0:

dagwieers replied:
Changed. I am going tired and don't even want to discuss anything anymore.

I'd rather get this pulled and then improved with individual PRs. I have rebased/merged conflicts 4 times now because fixes in this patch are being merged faster than this PR. So please merge this, and then offer PRs instead.

dagwieers committed:
Suggestion by @aperigault

dagwieers commented:
Please merge this one.

dagwieers committed:
Small documentation fixes

aperigault commented:
LGTM after shippable fix

aperigault commented:
@nerzhul @gundalow @jctanner Can you merge this PR ? There is a lot of important fixes. Thanks

nerzhul commented:
I need to test it before accept it, wait for 1 or 2 hours

nerzhul reviewed (approved):
ship_it
